src/Raven.Server/Documents/RetireAttachmentsSender.cs

```diff
@@ -0,0 +1,417 @@
+
+                                // get uploader for the attachment
+                                if (directUploaders.TryGetValue(doc.Id, out var directUpload) == false)
+                                {
+                                    // check if we have a destination configuration for the identifier
+
+                                    if (Configuration.Destinations.TryGetValue(doc.Id, out var destination) == false)
+                                    {
+                                        // no destination found for the identifier
+                                        if (Logger.IsWarnEnabled)
+                                            Logger.Warn($"No destination found for retired attachment with key: '{doc.LowerId}' and identifier '{doc.Id}'. Will try to retire it again!");
+
+                                        // this will keep the retired attachment in the tree and will try to retire it again afterwards.
+
+                                        totalCount--; // let's decrease the total count in case we have a lot of attachments without destination that we skip
+                                        continue;
+                                    }
+
+                                    if (destination.HasUploader() == false)
+                                    {
+                                        // no uploader for this destination
+                                        if (Logger.IsWarnEnabled)
+                                            Logger.Warn($"No uploader found for destination with identifier '{destination.Identifier}' for attachment with key: '{doc.LowerId}'. Will try to retire it again!");
+
+                                        totalCount--;
+                                        continue;
+                                    }
+
+                                    // create new uploader for the attachment
+                                    directUpload = new AttachmentUploader(UploaderSettings.GenerateDirectUploaderSettingsForAttachments(_database, doc.Id, destination.S3Settings, destination.AzureSettings, ConcurrentThreadsNumber), Logger, _token)
+                                    {
+                                        OnSuccess = x =>
+                                        {
+                                            totalUploaded += x.AttachmentSize;
+                                            retired.Enqueue(x.Doc);
+                                        },
+                                        OnException = x =>
+                                        {
+                                            var key = x.Doc.LowerId.ToString();
+                                            exceptions.Add(new UploadAttachmentException(x.Doc.Id, key, $"Upload task of retired attachment with key '{key}' and identifier '{x.Doc.Id}' failed.", x.UploadTask.Exception));
+                                        }
+                                    };
```

Comment:
this is super complex to follow - you have callback async processes that are called syncronously.
this looks like it remains from when you had threads.

Should be proper async API and be done with it.